chronos/osdefs.nim

```diff
@@ -0,0 +1,1562 @@
+
+  proc multiByteToWideChar*(codePage: uint32, dwFlags: uint32,
+                            lpMultiByteStr: ptr char, cbMultiByte: cint,
+                            lpWideCharStr: LPWSTR, cchWideChar: cint): cint {.
+       stdcall, dynlib: "kernel32", importc: "MultiByteToWideChar", sideEffect.}
+
+  proc getBestRouteXp*(dwDestAddr: uint32, dwSourceAddr: uint32,
+                       pBestRoute: ptr MibIpForwardRow): uint32 {.
+       stdcall, dynlib: "iphlpapi", importc: "GetBestRoute", sideEffect.}
+
+  proc queryPerformanceCounter*(res: var uint64) {.
+       stdcall, dynlib: "kernel32", importc: "QueryPerformanceCounter",
+       sideEffect.}
+
+  proc queryPerformanceFrequency*(res: var uint64) {.
+       stdcall, dynlib: "kernel32", importc: "QueryPerformanceFrequency",
+       sideEffect.}
+
+  proc getEnvironmentStringsW*(): LPWSTR {.
+       stdcall, dynlib: "kernel32", importc: "GetEnvironmentStringsW",
+       sideEffect.}
+
+  proc freeEnvironmentStringsW*(penv: LPWSTR): WINBOOL {.
+       stdcall, dynlib: "kernel32", importc: "FreeEnvironmentStringsW",
+       sideEffect.}
+
+  proc wcschr*(ws: LPWSTR, wc: WCHAR): LPWSTR {.
+       stdcall, dynlib: "ntdll", importc: "wcschr", sideEffect.}
+
+  proc getModuleHandle*(lpModuleName: WideCString): HANDLE {.
+       stdcall, dynlib: "kernel32", importc: "GetModuleHandleW", sideEffect.}
+
+  proc getProcAddress*(hModule: HANDLE, lpProcName: cstring): pointer {.
+       stdcall, dynlib: "kernel32", importc: "GetProcAddress", sideEffect.}
+
+  proc rtlNtStatusToDosError*(code: uint64): ULONG {.
+       stdcall, dynlib: "ntdll", importc: "RtlNtStatusToDosError", sideEffect.}
+
+  proc `==`*(x, y: SocketHandle): bool {.borrow.}
+  proc `==`*(x, y: HANDLE): bool {.borrow.}
+
+  proc getSecurityAttributes*(inheritHandle = false): SECURITY_ATTRIBUTES =
+    SECURITY_ATTRIBUTES(
+      nLength: DWORD(sizeof(SECURITY_ATTRIBUTES)),
+      lpSecurityDescriptor: nil,
+      bInheritHandle: if inheritHandle: TRUE else: FALSE
+    )
+
+elif defined(macos) or defined(macosx):
+  import std/posix
+  export posix
+
+  type
+    MachTimebaseInfo* {.importc: "struct mach_timebase_info",
+                        header: "<mach/mach_time.h>", pure, final.} = object
+      numer*: uint32
+      denom*: uint32
+
+  proc posix_gettimeofday*(tp: var Timeval, unused: pointer = nil) {.
+       importc: "gettimeofday", header: "<sys/time.h>".}
+
+  proc mach_timebase_info*(info: var MachTimebaseInfo) {.
+       importc, header: "<mach/mach_time.h>".}
+
+  proc mach_absolute_time*(): uint64 {.
+       importc, header: "<mach/mach_time.h>".}
+
+elif defined(linux):
+  import std/[posix]
+  export posix
+
+  when not defined(android) and defined(amd64):
+    const IP_MULTICAST_TTL*: cint = 33
+  else:
+    var IP_MULTICAST_TTL* {.importc: "IP_MULTICAST_TTL",
+                            header: "<netinet/in.h>".}: cint
+  const
+    EPOLLIN* = 0x00000001
```

Comment:
many of these constants are already (correctly) defined in upstream - why recreate them here? ie this one is part of `posix/epoll` for example ..

Reply:
There actually 3 issues.
1. Constants in `posix/epoll` and `posix/kqueue` are not get updated very often.
2. Continuous issues `Error: ambiguous identifier`.
3. We should monitor all the changes inside `epoll` and `kqueue` to avoid issues.

Reply:
1. well, if they are correct in upstream, this would be a benefit (they won't change them either)
2. should be solved with targeted imports/exports - ie the problem with ambiguous identifiers doesn't happen if we don't declare our own copy - this is exactly why it's better to use the common ones when possible, because code using chronos *and* upstream constants will work better (`import posix, chronos` breaks for example when both expose the same symbol, but doesn't break if chronos uses the one from posix)
3. chronos gets tested as part of nim testing, and we test chronos with multiple releases, so this is solved.

I think this has been more of a problem historically than it is now - we can merge this PR as-is but I think it's better to look for opportunities to use the upstream definitions when doing so is not outright wrong. One way to do it is basically to `from posix/epoll import all_correct_symbols` - this guards against future edits to `epoll` adding conflicting symbols while at the same time avoiding `ambiguous identifier` issues in downstream code.

Reply:
> I definitely do not want to use upstream for many reasons, there was already long history of updates which made broken every related package

This is not a sustainable option - but above all, it is also one that we have addressed, and having done that investment in upstream stability, we should now recognize the benefits and start profiting from them rather than holding on to an inefficient way of working around issues that no longer exist - ie we've addressed those problems together with upstream, and it holds specially true for "trivial" code like these constants - selectors is different because it introduces opinions whereas these are just trivial definitions. 

Copying things here perhaps makes it easier for chronos to implement itself maybe, but does a disservice to _users_ of chronos which now have to deal with ambiguous symbols, longer compile times and less interoperability.